// @tanstack/react-router/build/cjs/index.js
/**
 * @tanstack/react-router/src/index.tsx
 *
 * Copyright (c) TanStack
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

var history = require('@tanstack/history');
var invariant = require('tiny-invariant');
var warning = require('tiny-warning');
var awaited = require('./awaited.js');
var defer = require('./defer.js');
var CatchBoundary = require('./CatchBoundary.js');
var fileRoute = require('./fileRoute.js');
var lazyRouteComponent = require('./lazyRouteComponent.js');
var link = require('./link.js');
var Matches = require('./Matches.js');
var path = require('./path.js');
var qss = require('./qss.js');
var redirects = require('./redirects.js');
var route = require('./route.js');
var router = require('./router.js');
var RouterProvider = require('./RouterProvider.js');
var scrollRestoration = require('./scroll-restoration.js');
var searchParams = require('./searchParams.js');
var useBlocker = require('./useBlocker.js');
var useNavigate = require('./useNavigate.js');
var useParams = require('./useParams.js');
var useSearch = require('./useSearch.js');
var utils = require('./utils.js');



exports.invariant = invariant;
exports.warning = warning;
exports.Await = awaited.Await;
exports.useAwaited = awaited.useAwaited;
exports.defer = defer.defer;
exports.isDehydratedDeferred = defer.isDehydratedDeferred;
exports.CatchBoundary = CatchBoundary.CatchBoundary;
exports.CatchBoundaryImpl = CatchBoundary.CatchBoundaryImpl;
exports.ErrorComponent = CatchBoundary.ErrorComponent;
exports.FileRoute = fileRoute.FileRoute;
exports.lazyRouteComponent = lazyRouteComponent.lazyRouteComponent;
exports.Link = link.Link;
exports.useLinkProps = link.useLinkProps;
exports.Match = Matches.Match;
exports.MatchRoute = Matches.MatchRoute;
exports.Matches = Matches.Matches;
exports.Outlet = Matches.Outlet;
exports.matchContext = Matches.matchContext;
exports.useLoaderData = Matches.useLoaderData;
exports.useLoaderDeps = Matches.useLoaderDeps;
exports.useMatch = Matches.useMatch;
exports.useMatchRoute = Matches.useMatchRoute;
exports.useMatches = Matches.useMatches;
exports.useParentMatches = Matches.useParentMatches;
exports.cleanPath = path.cleanPath;
exports.interpolatePath = path.interpolatePath;
exports.joinPaths = path.joinPaths;
exports.matchByPath = path.matchByPath;
exports.matchPathname = path.matchPathname;
exports.parsePathname = path.parsePathname;
exports.removeBasepath = path.removeBasepath;
exports.resolvePath = path.resolvePath;
exports.trimPath = path.trimPath;
exports.trimPathLeft = path.trimPathLeft;
exports.trimPathRight = path.trimPathRight;
exports.decode = qss.decode;
exports.encode = qss.encode;
exports.isRedirect = redirects.isRedirect;
exports.redirect = redirects.redirect;
exports.NotFoundRoute = route.NotFoundRoute;
exports.RootRoute = route.RootRoute;
exports.Route = route.Route;
exports.RouteApi = route.RouteApi;
exports.createRouteMask = route.createRouteMask;
exports.rootRouteId = route.rootRouteId;
exports.rootRouteWithContext = route.rootRouteWithContext;
exports.PathParamError = router.PathParamError;
exports.Router = router.Router;
exports.SearchParamError = router.SearchParamError;
exports.componentTypes = router.componentTypes;
exports.getInitialRouterState = router.getInitialRouterState;
exports.lazyFn = router.lazyFn;
exports.RouterProvider = RouterProvider.RouterProvider;
exports.getRouteMatch = RouterProvider.getRouteMatch;
Object.defineProperty(exports, 'routerContext', {
	enumerable: true,
	get: function () { return RouterProvider.routerContext; }
});
exports.useRouter = RouterProvider.useRouter;
exports.useRouterState = RouterProvider.useRouterState;
exports.ScrollRestoration = scrollRestoration.ScrollRestoration;
exports.useElementScrollRestoration = scrollRestoration.useElementScrollRestoration;
exports.useScrollRestoration = scrollRestoration.useScrollRestoration;
exports.defaultParseSearch = searchParams.defaultParseSearch;
exports.defaultStringifySearch = searchParams.defaultStringifySearch;
exports.parseSearchWith = searchParams.parseSearchWith;
exports.stringifySearchWith = searchParams.stringifySearchWith;
exports.Block = useBlocker.Block;
exports.useBlocker = useBlocker.useBlocker;
exports.Navigate = useNavigate.Navigate;
exports.useNavigate = useNavigate.useNavigate;
exports.useParams = useParams.useParams;
exports.useSearch = useSearch.useSearch;
exports.deepEqual = utils.deepEqual;
exports.escapeJSON = utils.escapeJSON;
exports.functionalUpdate = utils.functionalUpdate;
exports.isPlainObject = utils.isPlainObject;
exports.isServer = utils.isServer;
exports.last = utils.last;
exports.pick = utils.pick;
exports.replaceEqualDeep = utils.replaceEqualDeep;
exports.shallow = utils.shallow;
exports.useLayoutEffect = utils.useLayoutEffect;
exports.useRouteContext = utils.useRouteContext;
exports.useStableCallback = utils.useStableCallback;
Object.keys(history).forEach(function (k) {
	if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
		enumerable: true,
		get: function () { return history[k]; }
	});
});
